### PR TITLE
split out embed/thread plugins and add PHP plugin

### PR DESCRIPTION
**Pull request by jwestbrook**
Per discussion on #172 I've changed the plugin loading to determine if the plugin is loading from embedded or threaded folders and load them accordingly.

Then I added the php plugin I wrote to take advantage of the node connection and run `php -l` from the command line based on `STDIN` and report the results.

Some problems I've ran into
- if the php code you are running sets the error level low or off you only get that the parse error exists and what line
- you only get parse errors as that's what `php -l` reports (no warnings/deprecated etc)

Todo
- I need to add config options so that Windows users can point to their php binary or users that don't have php in $PATH

### CODE REVIEW

File: pluginLoader.js [hunk 1]
```diff
@@ -33,7 +33,12 @@ define(function(require, exports, module){
             var _lint = plugin.lint;
 
             function lint(text, settings) {
-                return Promise.resolve(_lint(text, settings));
+                var results = _lint(text, settings,pluginsMeta);
+                if(results.promise !== undefined && results.promise !== null){
```
Comment (MiguelCastillo) [inline pluginLoader.js]: I couldn't find where result gets a promise added as a property...  Why is that needed?

**Comment by MiguelCastillo** [comment]
This is really cool!!! :D

File: pluginLoader.js (hunk 1 above)
Comment (jwestbrook) [inline pluginLoader.js]: So I'm checking to see if the `lint()` method returns a promise or a standard object.

At line 15 in plugins/default/embedded/php/main.js I setup a Promise to be returned - the Promise object has a `promise` property /shrug it should probably be checking in a better way I guess.

I needed to use a promise as the `child_process.exec()` method is async - so the results of the process would not be available when the `lint()` method @ line 7 returns. So then I followed the data flow (which crossed my eyes a few times) and at this point ^^ you are resolving the Promise with the lint results from whichever plugin is returning. So instead I passed the Promise through.

does that explain it?

**Comment by RedDude** [comment]
This will never be accepted?

**Comment by MiguelCastillo** [comment]
Yes! :)

**Comment by MiguelCastillo** [comment]
@jwestbrook there are some tweaks that should be made to the implementation... I merged it as is. and we can make the tweaks in separate PRs.  Any interest in being added as a contributor?

File: pluginLoader.js (hunk 1 above)
Comment (MiguelCastillo) [inline pluginLoader.js]: @jwestbrook apologies for the terribly slow response.

Promise.resolve will gracefully convert input objects to promises... And input promises are basically a noop.  That's a good way to uniformly handle promises and non promises objects consistently.

``` javascript
return Promise.resolve(_lint(text, settings, pluginsMeta));
```

Should be sufficient.

**Comment by jwestbrook** [comment]
@MiguelCastillo sure
